samples/core/multi-device/main.c

```diff
@@ -0,0 +1,702 @@
+                      const cl_uint x_dim, const cl_uint y_dim)
+{
+    const cl_uint mask_dim = 3;
+    const cl_uint pad_width = mask_dim / 2;
+    const cl_uint in_dim_x = x_dim + pad_width * 2;
+
+    for (cl_uint gid_x = 0; gid_x < x_dim; ++gid_x)
+    {
+        for (cl_uint gid_y = 0; gid_y < y_dim; ++gid_y)
+        {
+            float result = 0.f;
+            for (cl_uint y = 0; y < mask_dim; ++y)
+            {
+                for (cl_uint x = 0; x < mask_dim; ++x)
+                {
+                    result += mask[y * mask_dim + x]
+                        * in[(gid_y + y) * in_dim_x + (gid_x + x)];
+                }
+            }
+            out[gid_y * x_dim + gid_x] = result;
+        }
+    }
+}
+
+cl_int opencl_version_contains(const cl_device_id dev,
+                               const char* version_fragment)
+{
+    char version[64];
+    cl_int error = CL_SUCCESS;
+    OCLERROR_RET(clGetDeviceInfo(dev, CL_DEVICE_VERSION, sizeof(version),
+                                 &version, NULL),
+                 error, fail);
+    char* found_version = strstr(version, version_fragment);
+    return (found_version != NULL);
+fail:
+    return error;
+}
+
+int main(int argc, char* argv[])
+{
+    cl_int error = CL_SUCCESS;
+    cl_int end_error = CL_SUCCESS;
+    cl_device_id dev;
+    cl_context context;
+    cl_program program;
+    cl_mem dev_input_grid, dev_output_grid, dev_mask;
+
+    cl_kernel convolutions[2] = { 0 };
+    cl_command_queue sub_queues[2] = { 0 };
+    cl_mem sub_input_grids[2] = { 0 };
+    cl_mem sub_output_grids[2] = { 0 };
+    cl_event events[2] = { 0 };
+
+    // Parse command-line options.
+    struct cl_sdk_options_Diagnostic diag_opts = { .quiet = false,
+                                                   .verbose = false };
+
+    // By default assume that there is only one device available.
+    // dev_opts->number is set to 1 so that when calling to cl_util_get_device
+    // for the second device there is no index out of range.
+    struct cl_sdk_options_SingleDevice dev_opts = {
+        .triplet = { 0, 0, CL_DEVICE_TYPE_ALL }
+    };
+    struct convolution_options convolution_opts = { .x_dim = 4096,
+                                                    .y_dim = 4096 };
+
+    OCLERROR_RET(
+        parse_options(argc, argv, &diag_opts, &dev_opts, &convolution_opts),
+        error, end);
+
+    // Create runtime objects based on user preference or default.
+    OCLERROR_PAR(dev = cl_util_get_device(dev_opts.triplet.plat_index,
+                                          dev_opts.triplet.dev_index,
+                                          dev_opts.triplet.dev_type, &error),
+                 error, end);
+
+    if (!diag_opts.quiet)
+    {
+        cl_util_print_device_info(dev);
+    }
+
+    if (diag_opts.verbose)
+    {
+        printf("Creating sub-devices...");
+        fflush(stdout);
+    }
+
+    if (opencl_version_contains(dev, "1.1"))
```

Comment:
Nitpick: this should be checking for OpenCL 1.0 and 1.1, not just OpenCL 1.1.

Reply:
Added check for 1.0 (also in the c++ sample)!